File: hs/Mod_csmAlg.hs
module Mod_csmAlg ( csmAlg ) where
import Prelude hiding (and, or) 
import Debug.Trace
import Splib
import Types

csmAlg :: CSMInstance -> Bool
csmAlg x | coa x == 0 && cob x == 0 = True
         | coa x == 0 || cob x == 0 = False
		 | otherwise = csmAlgHelper x 0 0 0 0

csmAlgHelper :: CSMInstance -> Int -> Int -> Int -> Int -> Bool
csmAlgHelper x ia ib acta actb | ia == coa x && ib == cob x = True
                             | sp_not (ssc x ia ib acta actb minl) = False
                             | ia < coa x && ib < cob x = nextStep x ia ib (acta+minl) (actb+minl)
                             | ia < coa x || ia < cob x = False
                              where 
                                 la = lena x (sp_min ia ((coa x)-1))
                                 lb = lenb x (sp_min ib ((cob x)-1))
                                 minl = sp_min (la-acta) (lb-actb)

nextStep :: CSMInstance -> Int -> Int -> Int -> Int -> Bool
nextStep x ia ib acta actb | acta == la && actb == lb = csmAlgHelper x (ia+1) (ib+1) 0 0
                         | acta == la = csmAlgHelper x (ia+1) ib 0 actb
						 | actb == lb = csmAlgHelper x ia (ib+1) acta 0
						  where
                             la = lena x ia
                             lb = lenb x ib

File: cpp/Mod_csmAlg.cpp
#pragma once

bool csmAlg(CSMInstance x){
	int ia = 0;
	int ib = 0;
	
	if(coa(x) == 0 && cob(x) == 0)
		return true;
	if(coa(x) == 0 || cob(x) == 0)
		return false;
	
	int la = lena(x, 0);
	int lb = lenb(x, 0);
	
	int acta = 0;
	int actb = 0;
	
	int minl = sp_min(la,lb);
	
	while(ia < coa(x) && ib < cob(x)){
		if(!ssc(x, ia, ib, acta, actb, minl)){
			return false;
		}
		
		acta += minl;
		actb += minl;
		
		if(acta == la){
			ia++;
			la = lena(x, ia);
			acta = 0;
		}
        
		if(actb == lb){
			ib++;
			lb = lenb(x, ib);
			actb = 0;
		}
		
		minl = sp_min(la-acta, lb-actb);
	}
	
	if(ia < coa(x) || ib < cob(x))
		return false;
	
	return true;
}
